presto-hive/src/test/java/com/facebook/presto/hive/TestEventListenerWithExchangeMaterialization.java

```diff
@@ -101,58 +103,67 @@ public void testRuntimeOptimizedStagesCorrectness()
 
-        assertEquals(queryMetadata.getRuntimeOptimizedStages().size(), 1);
-        assertEquals(queryMetadata.getRuntimeOptimizedStages().get(0), "1");
-        assertTrue(runtimeOptimizedStages.isPresent());
-        assertEquals(runtimeOptimizedStages.get().size(), 1);
-        assertEquals(queryMetadata.getRuntimeOptimizedStages(), runtimeOptimizedStages.get().stream()
-                .map(stageId -> String.valueOf(stageId.getId()))
-                .collect(toImmutableList()));
+        if (queryMetadata.getQueryId().equals(queryId.toString())) {
+            assertEquals(queryMetadata.getRuntimeOptimizedStages().size(), 1);
+            assertEquals(queryMetadata.getRuntimeOptimizedStages().get(0), "1");
+            assertTrue(runtimeOptimizedStages.isPresent());
+            assertEquals(runtimeOptimizedStages.get().size(), 1);
+            assertEquals(queryMetadata.getRuntimeOptimizedStages(), runtimeOptimizedStages.get().stream()
+                    .map(stageId -> String.valueOf(stageId.getId()))
+                    .collect(toImmutableList()));
+        }
 
         // Now, the following query should have two runtime optimized stages: 1 and 4, corresponding to the two join operations (on regionkey and nationkey respectively).
-        runQueryAndWaitForEvents("WITH natreg AS (SELECT nation.regionkey, nationkey, region.name FROM region INNER JOIN nation ON nation.regionkey=region.regionkey) SELECT phone, regionkey FROM natreg INNER JOIN supplier ON supplier.nationkey=natreg.nationkey", expectedEvents);
+        queryId = runQueryAndWaitForEvents("WITH natreg AS (SELECT nation.regionkey, nationkey, region.name FROM region INNER JOIN nation ON nation.regionkey=region.regionkey) SELECT phone, regionkey FROM natreg INNER JOIN supplier ON supplier.nationkey=natreg.nationkey", expectedEvents);
         queryCreatedEvent = getOnlyElement(generatedEvents.getQueryCreatedEvents());
         queryCompletedEvent = getOnlyElement(generatedEvents.getQueryCompletedEvents());
         queryMetadata = queryCompletedEvent.getMetadata();
         runtimeOptimizedStages = queryRunner.getCoordinator().getQueryManager().getFullQueryInfo(new QueryId(queryCreatedEvent.getMetadata().getQueryId())).getRuntimeOptimizedStages();
 
-        assertEquals(queryMetadata.getRuntimeOptimizedStages().size(), 2);
-        assertEquals(ImmutableSet.copyOf(queryMetadata.getRuntimeOptimizedStages()), ImmutableSet.of("1", "4"));
-        assertTrue(runtimeOptimizedStages.isPresent());
-        assertEquals(runtimeOptimizedStages.get().size(), 2);
-        assertEquals(queryMetadata.getRuntimeOptimizedStages(), runtimeOptimizedStages.get().stream()
-                .map(stageId -> String.valueOf(stageId.getId()))
-                .collect(toImmutableList()));
+        if (queryMetadata.getQueryId().equals(queryId.toString())) {
```

Comment:
I worry that this means at some point the test can break and we won't know it because we aren't doing any checks. Is this part necessary for fixing the flakiness?  Is there a way to wait for the correct event instead?

Reply:
No, this part is not essential, just for the super extreme case where after waiting 180 seconds, the previous queryCompletedEvent still have not reached.

Reply:
great, can you remove this then? If there's still test flakiness after this, I'd rather add a loop to wait for the correct query id for a certain amount of time.

Reply:
But on the other hand, if the queryIds are different then there is no point to further run assertions because essentially we will be comparing two events from two different queries.

Reply:
Sure. I will remove this part.